lib/src/core/models/functions/functions.dart

```diff
@@ -0,0 +1,120 @@
+import 'dart:convert';
+
+import 'package:meta/meta.dart';
+
+@immutable
+class OpenAiFunctionModel {
+  final String name;
+  final String? description;
+  final OpenAiFunctionParameters? parameters;
+
+  const OpenAiFunctionModel({
+    required this.name,
+    this.description,
+    this.parameters,
+  });
+
+  Map<String, dynamic> toMap() {
+    return {
+      'name': name,
+      if (description != null) 'description': description,
+      if (parameters != null) 'parameters': parameters!.toMap(),
+    };
+  }
+}
+
+class OpenAiFunctionParameters {
+  final String type;
+  final List<OpenAiFunctionProperty> properties;
+  OpenAiFunctionParameters({
+    this.type = 'object',
+    required this.properties,
+  });
+
+  Map<String, dynamic> toMap() {
+    final requiredProperties = properties
+        .where((property) => property.isRequired)
+        .map((property) => property.name)
+        .toList();
+    return {
+      'type': type,
+      'properties': {
+        for (final parameter in properties)
```

Comment:
can this be moved to a separated function to generate the `properties` field?

Reply:
Done with the refactor